analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -63,21 +77,102 @@ public override void ExecutionCompleted()
+        {
+            ReportIssue(Rule2583, operation, secondaryLocations, string.Format(MessageUnreachable, issueMessage));
+        }
+        else
+        {
+            ReportIssue(Rule2589, operation, null, string.Format(issueMessage, conditionEvaluation));
+        }
     }
+
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionIsTrue)
+    {
+        var unreachable = Unreachable(block, conditionIsTrue).Where(x => IsNotIgnored(x.Syntax));
+        List<Location> locations = new();
+        IOperation currentStart = null;
+        IOperation currentEnd = null;
+        foreach (var operation in unreachable.Union(reached).OrderBy(x => x.Syntax.SpanStart))
+        {
+            if (reached.Contains(operation))
+            {
+                if (currentStart is not null)
+                {
+                    locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+                    currentStart = null;
+                }
+            }
+            else
+            {
+                currentStart ??= operation;
+                currentEnd = operation;
+            }
+        }
+        if (currentStart != null)
+        {
+            locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+        }
+        return locations;
+    }
+
+    private static IEnumerable<IOperation> Unreachable(BasicBlock block, bool conditionIsTrue)
+    {
+        if (block.SuccessorBlocks.Distinct().Count() != 2)
+        {
+            return Enumerable.Empty<IOperation>();
+        }
+        HashSet<BasicBlock> reachable = new() { block };
+        HashSet<BasicBlock> unreachable = new();
+        var successor = conditionIsTrue ^ block.ConditionKind == ControlFlowConditionKind.WhenFalse
+            ? block.ConditionalSuccessor.Destination
+            : block.SuccessorBlocks.Single(x => x != block.ConditionalSuccessor.Destination);
+        TraverseReachable(successor);
+        var unreachableSuccessor = block.SuccessorBlocks.Single(x => x != successor);
+        TraverseUnreachable(unreachableSuccessor);
+        return unreachable.SelectMany(x => x.OperationsAndBranchValue);
+
+        void TraverseReachable(BasicBlock block)
+        {
+            if (reachable.Add(block))
+            {
+                foreach (var successor in block.SuccessorBlocks)
+                {
+                    TraverseReachable(successor);
+                }
+            }
+        }
+
+        void TraverseUnreachable(BasicBlock block)
+        {
+            if (reachable.Contains(block) is false && unreachable.Add(block))
+            {
+                foreach (var successor in block.SuccessorBlocks)
+                {
+                    TraverseUnreachable(successor);
+                }
+            }
+        }
+    }
+
+    private bool IsNotIgnored(SyntaxNode x) =>
+        !IsForLoopIncrementor(x)
+        && !IsConditionalAccessExpression(x)
+        && !IsLeftCoalesceExpression(x);
```

Comment:
```suggestion
    private bool IsIgnoredLocation(SyntaxNode x) =>
        IsForLoopIncrementor(x)
        || IsConditionalAccessExpression(x)
        || IsLeftCoalesceExpression(x);
```

This should also be a local function under `ReportIssue(IOperation operation, BasicBlock block, bool conditionEvaluation)` since it's used only there.

Reply:
Changed the name. Not making it a local as discussed.